File: qiskit/synthesis/multi_controlled/mcx.py
# This code is part of Qiskit.
#
# (C) Copyright IBM 2024.
#
# This code is licensed under the Apache License, Version 2.0. You may
# obtain a copy of this license in the LICENSE.txt file in the root directory
# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
#
# Any modifications or derivative works of this code must retain this
# copyright notice, and modified files need to carry a notice indicating
# that they have been altered from the originals.

"""Synthesis algorithms for MCX gates."""
from __future__ import annotations

from math import ceil
import numpy as np

from qiskit.circuit.quantumregister import QuantumRegister
from qiskit.circuit.quantumcircuit import QuantumCircuit
from qiskit.circuit.library import CXGate, CCXGate, C3XGate, C4XGate


def synth_mcx_recursive(num_control_qubits: int, num_clean_ancillas: int) -> QuantumCircuit | None:
    """
    Synthesize MCX gate using the recursive algorithm used for ``MCXRecursive`` gates.

    This method does not require any ancilla qubits when the number of control qubits is
    up to 4, and requires exactly one ancilla qubits when the number of control qubits is
    greater than or equal to 5.

    Args:
        num_control_qubits: The number of control qubits for the gate.
        num_ancilla_qubits: The number of available ancilla qubits.

    Returns:
        The synthesized quantum circuit, or ``None`` if the number of available ancilla qubits
            is not sufficient.

    """

    print(f"In synth_mcx_recursive: {num_clean_ancillas = }")

    if num_control_qubits <= 4:
        num_clean_ancillas_required = 0
    else:
        num_clean_ancillas_required = 1

    if num_clean_ancillas < num_clean_ancillas_required:
        # Not enough ancilla qubits available to run the algorithm
        return None

    num_qubits = num_control_qubits + num_clean_ancillas_required + 1

    q = QuantumRegister(num_qubits, name="q")
    qc = QuantumCircuit(q)
    if num_control_qubits == 1:
        qc._append(CXGate(), q[:], [])
    elif num_control_qubits == 2:
        qc._append(CCXGate(), q[:], [])
    elif num_control_qubits == 3:
        qc._append(C3XGate(), q[:], [])
    elif num_control_qubits == 4:
        qc._append(C4XGate(), q[:], [])
    else:
        for instr, qargs, cargs in _recurse(q[:-1], q_ancilla=q[-1]):
            qc._append(instr, qargs, cargs)
    return qc


def _recurse(q, q_ancilla=None):
    # print(f"_RECURSE: {q = }, {q_ancilla = }")
    # recursion stop
    if len(q) == 4:
        return [(C3XGate(), q[:], [])]
    if len(q) == 5:
        return [(C4XGate(), q[:], [])]
    if len(q) < 4:
        raise AttributeError("Something went wrong in the recursion, have less than 4 qubits.")

    # recurse
    num_ctrl_qubits = len(q) - 1
    middle = ceil(num_ctrl_qubits / 2)
    first_half = [*q[:middle], q_ancilla]
    second_half = [*q[middle:num_ctrl_qubits], q_ancilla, q[num_ctrl_qubits]]

    rule = []
    rule += _recurse(first_half, q_ancilla=q[middle])
    rule += _recurse(second_half, q_ancilla=q[middle - 1])
    rule += _recurse(first_half, q_ancilla=q[middle])
    rule += _recurse(second_half, q_ancilla=q[middle - 1])

    return rule


def synth_mcx_using_mcphase(num_control_qubits: int) -> QuantumCircuit | None:
    """
    Synthesize MCX gate using the reduction to MCPhase.

    This method does not require any ancilla qubits.

    Args:
        num_control_qubits: The number of control qubits for the gate.

    Returns:
        The synthesized quantum circuit.
    """
    num_qubits = num_control_qubits + 1

    q = QuantumRegister(num_qubits, name="q")
    qc = QuantumCircuit(q)

    if num_control_qubits == 1:
        qc._append(CXGate(), q[:], [])
    elif num_control_qubits == 2:
        qc._append(CCXGate(), q[:], [])
    elif num_control_qubits == 3:
        qc._append(C3XGate(), q[:], [])
    elif num_control_qubits == 4:
        qc._append(C4XGate(), q[:], [])
    else:
        q_controls = list(range(num_control_qubits))
        q_target = num_control_qubits
        qc.h(q_target)
        qc.mcp(np.pi, q_controls, q_target)
        qc.h(q_target)

    return qc

File: crates/accelerate/src/synthesis/multi_controlled/mcx.rs
// This code is part of Qiskit.
//
// (C) Copyright IBM 2025
//
// This code is licensed under the Apache License, Version 2.0. You may
// obtain a copy of this license in the LICENSE.txt file in the root directory
// of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
//
// Any modifications or derivative works of this code must retain this
// copyright notice, and modified files need to carry a notice indicating
// that they have been altered from the originals.

use pyo3::types::PyAnyMethods;
use pyo3::{PyResult, Python};
use qiskit_circuit::imports;
use qiskit_circuit::operations::{Operation, PyGate, StandardGate};
use qiskit_circuit::{circuit_data::CircuitData, operations::Param, Qubit};

use std::f64::consts::PI;
const PI2: f64 = PI / 2.0;

/// Efficient synthesis for CCX.
pub fn ccx() -> PyResult<CircuitData> {
    let circuit = StandardGate::CCX
        .definition(&[])
        .expect("Could not extract definition for CCX.");
    Ok(circuit)
}

/// Efficient synthesis for RCCX.
fn rccx() -> PyResult<CircuitData> {
    let circuit = StandardGate::RCCX
        .definition(&[])
        .expect("Could not extract definition for RCCX.");
    Ok(circuit)
}

/// Efficient synthesis for 3-controlled X-gate.
pub fn c3x() -> PyResult<CircuitData> {
    let circuit = StandardGate::C3X
        .definition(&[])
        .expect("Could not extract definition for C3X.");
    Ok(circuit)
}

/// Efficient synthesis for 4-controlled X-gate.
pub fn c4x() -> PyResult<CircuitData> {
    let mut circuit = CircuitData::with_capacity(5, 0, 0, Param::Float(0.0))?;
    circuit.h(4);
    circuit.cp(PI2, 3, 4);
    circuit.h(4);
    circuit.compose(
        &CircuitData::from_standard_gate_definition(StandardGate::RC3X, &[])?,
        Some(&[Qubit(0), Qubit(1), Qubit(2), Qubit(3)]),
    )?;
    circuit.h(4);
    circuit.cp(-PI2, 3, 4);
    circuit.h(4);
    circuit.compose(
        &CircuitData::from_standard_gate_definition(StandardGate::RC3X, &[])?.inverse()?,
        Some(&[Qubit(0), Qubit(1), Qubit(2), Qubit(3)]),
    )?;
    circuit.compose(
        &CircuitData::from_standard_gate_definition(StandardGate::C3SX, &[])?,
        Some(&[Qubit(0), Qubit(1), Qubit(2), Qubit(4)]),
    )?;
    Ok(circuit)
}

/// A block in the `action part`, see Iten et al.
fn action_gadget() -> PyResult<CircuitData> {
    let mut circuit = CircuitData::with_capacity(3, 0, 5, Param::Float(0.0))?;
    circuit.h(2);
    circuit.t(2);
    circuit.cx(0, 2);
    circuit.tdg(2);
    circuit.cx(1, 2);
    Ok(circuit)
}

/// A block in the `reset part`, see Iten et al.
fn reset_gadget() -> PyResult<CircuitData> {
    let mut circuit = CircuitData::with_capacity(3, 0, 5, Param::Float(0.0))?;
    circuit.cx(1, 2);
    circuit.t(2);
    circuit.cx(0, 2);
    circuit.tdg(2);
    circuit.h(2);
    Ok(circuit)
}

/// Synthesize a multi-controlled X gate with :math:`k` controls based on the paper
/// by Iten et al. [1].
///
/// For :math:`k\ge 4` the method uses :math:`k - 2` dirty ancillary qubits, producing a circuit
/// with :math:`2 * k - 1` qubits and at most :math:`8 * k - 6` CX gates. For :math:`k\le 3`
/// explicit efficient circuits are used instead.
///
/// # Arguments
/// - num_controls: the number of control qubits.
/// - relative_phase: when set to `true`, the method applies the optimized multi-controlled
///   X gate up to a relative phase, in a way that the relative phases of the `action part`
///   cancel out with the relative phases of the `reset part`.
/// - action_only: when set to `true`, the methods applies only the `action part`.
///
/// # References
///
/// 1. Iten et al., *Quantum Circuits for Isometries*, Phys. Rev. A 93, 032318 (2016),
/// [arXiv:1501.06911] (http://arxiv.org/abs/1501.06911).
pub fn synth_mcx_n_dirty_i15(
    num_controls: usize,
    relative_phase: bool,
    action_only: bool,
) -> PyResult<CircuitData> {
    if num_controls == 1 {
        let mut circuit = CircuitData::with_capacity(2, 0, 1, Param::Float(0.0))?;
        circuit.cx(0, 1);
        Ok(circuit)
    } else if num_controls == 2 {
        ccx()
    } else if num_controls == 3 && !relative_phase {
        c3x()
    } else {
        let num_ancillas = num_controls - 2;
        let num_qubits = num_controls + 1 + num_ancillas;
        let mut circuit = CircuitData::with_capacity(num_qubits as u32, 0, 0, Param::Float(0.0))?;

        let controls: Vec<u32> = (0..num_controls).map(|q| q as u32).collect();
        let target = num_controls as u32;
        let ancillas: Vec<u32> = ((num_controls + 1)..num_qubits).map(|q| q as u32).collect();

        for j in 0..2 {
            if !relative_phase {
                circuit.compose(
                    &ccx()?,
                    Some(&[
                        Qubit(controls[num_controls - 1]),
                        Qubit(ancillas[num_controls - 3]),
                        Qubit(target),
                    ]),
                )?;
            } else if j == 0 {
                circuit.compose(
                    &action_gadget()?,
                    Some(&[
                        Qubit(controls[num_controls - 1]),
                        Qubit(ancillas[num_controls - 3]),
                        Qubit(target),
                    ]),
                )?;
            } else if j == 1 {
                circuit.compose(
                    &reset_gadget()?,
                    Some(&[
                        Qubit(controls[num_controls - 1]),
                        Qubit(ancillas[num_controls - 3]),
                        Qubit(target),
                    ]),
                )?;
            }

            // action part
            for i in (0..num_controls - 3).rev() {
                circuit.compose(
                    &action_gadget()?,
                    Some(&[
                        Qubit(controls[i + 2]),
                        Qubit(ancillas[i]),
                        Qubit(ancillas[i + 1]),
                    ]),
                )?;
            }

            circuit.compose(
                &rccx()?,
                Some(&[Qubit(controls[0]), Qubit(controls[1]), Qubit(ancillas[0])]),
            )?;

            // reset part
            for i in 0..num_controls - 3 {
                circuit.compose(
                    &reset_gadget()?,
                    Some(&[
                        Qubit(controls[i + 2]),
                        Qubit(ancillas[i]),
                        Qubit(ancillas[i + 1]),
                    ]),
                )?;
            }

            if action_only {
                circuit.compose(
                    &ccx()?,
                    Some(&[
                        Qubit(controls[num_controls - 1]),
                        Qubit(ancillas[num_controls - 3]),
                        Qubit(target),
                    ]),
                )?;
                break;
            }
        }
        Ok(circuit)
    }
}

/// Synthesize a multi-controlled X gate with :math:`k` controls based on
/// the implementation for `MCPhaseGate`.
///
/// In turn, the MCPhase gate uses the decomposition for multi-controlled
/// special unitaries described in [1].
///
/// # Arguments
/// - num_controls: the number of control qubits.
///
/// # Returns
///
/// A quantum circuit with :math:`k + 1` qubits. The number of CX-gates is
/// quadratic in :math:`k`.
///
/// # References
///
/// 1. Vale et. al., *Circuit Decomposition of Multicontrolled Special Unitary
/// Single-Qubit Gates*, IEEE TCAD 43(3) (2024),
/// [arXiv:2302.06377] (https://arxiv.org/abs/2302.06377).
pub fn synth_mcx_noaux_v24(py: Python, num_controls: usize) -> PyResult<CircuitData> {
    println!("IN RUST synth_mcx_noaux_v24! with num_controls = {}", num_controls);
    if num_controls == 3 {
        c3x()
    } else if num_controls == 4 {
        c4x()
    } else {
        let num_qubits = (num_controls + 1) as u32;
        let target = num_controls as u32;

        let mut circuit = CircuitData::with_capacity(num_qubits, 0, 0, Param::Float(0.0))?;
        circuit.h(target);

        let mcphase_cls = imports::MCPHASE_GATE.get_bound(py);
        let mcphase_gate = mcphase_cls
            .call1((PI, num_controls))
            .expect("Could not create MCPhaseGate Python-side");

        let as_py_gate = PyGate {
            qubits: num_qubits,
            clbits: 0,
            params: 1,
            op_name: "mcphase".to_string(),
            gate: mcphase_gate.into(),
        };

        circuit.push_py_gate(
            as_py_gate,
            &[],
            &(0..num_qubits).map(Qubit).collect::<Vec<Qubit>>(),
            &[],
        )?;

        circuit.h(target);

        Ok(circuit)
    }
}
